### ISSUE TITLE
X/Y Plot working with single-value PVs

### ISSUE THREAD

**Issue by claudio-rosati** [kasemir/org.csstudio.display.builder#391]
My integrators are asking a better compatibility of X/Y Plot with BOY's X/Y Graph.

Being that I've also request for a polar plot, I'm not sure if:

1. Upgrade X/Y Plot to accept single-value PVs and implement a new Polar Plot widget, or
2. Implement a new Plot widget (Multi Plot?) to perform X/Y and Polar drawings with single-value PVs.

@kasemir 
What do you think?

In case of 1., will you do the change, or point me in the right sources to do it myself?

**Comment by claudio-rosati** [comment]
So, if I've understood correctly,

is not possible to have a plot of Y over X if they are scalars.

So how do we want to allow this? I'm fine to to do this kind of widget:

- X/Y (both arrays) -> X/Y Plot (existing);
- X/time (X scalar or array) -> Data Browser (existing);
- X/Y and Rho/Phi (both scalar) -> Scalar Plot (to be done).

To solve mixed cases, maybe we can provide some loc/sim PVs to convert an array PV into a scalar one (read-only) and vice versa.

Do you agree on that?

**Comment by kasemir** [comment]
Exactly. A scalar Y against a scalar X is fundamentally different from waveform (XYPlot) as well as scalar-over-time (data browser), so it needs to be a new type of plot, like a "Scalar Scatter Plot" that you plan to implement.

It's tricky, though.
When two scalar PVs X and Y update, you get two separate notifications.
I think BOY simply added received data to an in-client circular buffer for each channel.
With that, it's easy to be one step out of sync.
Assume X and Y send updates x0, y0, x1, y1, x2, y2, x3, y3, x4, y4, ... at 10Hz.
BOY starts, and the first values it receives happen to be  y1, x2, y2, x3, y3, x4, y4, ...
So it now has in-client arrays Xhist = [x2, x3, x4, ...] and Yhist = [ y1, y2, y3, .. ] from which it happily plots (x2, y1), (x3, y2), ..

How do you intent to prevent his?
Do you then check their time stamps to see if these describe the same point (X, Y) on the plot?

To handle that right now in a more dependably way, have the IOC that has the X and Y also produce X and Y waveforms based on the compress record.
When the IOC processes X and Y, it can at that time FLNK to an Xhist and Yhist compress record with ALG="Circular Buffer", N=(size of desired waveform).
So now the IOC produces correct Xhist and Yhist arrays.
Then use the existing XYPlot for Xhist and Yhist -> done.

Sorry, but BOY's X/Y Graph had problems. People not understanding it and asking for the bugs back is not a good idea.

**Comment by claudio-rosati** [comment]
I understand, and I think that the only way is using the timestamp and possibly (or, better, optionally) a linear interpolator to get intermediate values. It is something I've already done in the past.
